ambry-store/src/main/java/com/github/ambry/store/StorageManager.java

```diff
@@ -817,4 +826,165 @@ private void resumeDecommission(String partitionName) throws Exception {
+    // All the failed unavailable disks are the failed disks
+    private final List<DiskId> failedDisk = diskToDiskManager.keySet()
+        .stream()
+        .filter(diskId -> diskId.getState() == HardwareState.UNAVAILABLE)
+        .collect(Collectors.toList());
+    private final long acquireLockBackoffTime = storeConfig.storeDiskFailureHandlerRetryLockBackoffTimeInSeconds * 1000;
+    private final int capacityReportingPercentage = storeConfig.storeDiskCapacityReportingPercentage;
+
+    @Override
+    public void run() {
+      if (!clusterMap.isDataNodeInFullAutoMode(currentNode)) {
+        return;
+      }
+      logger.info("Current Node is in FULL_AUTO, try to detect disk failure.");
+      // First, we have to detect if there is a new disk failure
+      List<DiskId> newFailedDisks = diskToDiskManager.keySet()
+          .stream()
+          .filter(diskId -> !isDiskAvailable(diskId) && !failedDisk.contains(diskId))
+          .collect(Collectors.toList());
+      if (newFailedDisks.isEmpty()) {
+        return;
+      }
+      logger.info("Failed disk detected: {}", newFailedDisks);
+      storeMainMetrics.handleDiskFailureCount.inc();
+
+      // When there is a new disk failure, we need to do several things
+      // 1. remove replicasOnFailedDisks from the property store
+      // 2. reset the partitions
+      // 3. update the capacity to instance config
+      // When reset the partitions, we don't expect Helix sending downward state transition messages to this host, to
+      // transition those replicasOnFailedDisks from LEADER/STANDBY all the way down to DROPPED. Even if helix does send state
+      // transition messages, we won't be able to do anything since the disk is not healthy. In this case, we have to
+      // remove the replica from property store right away.
+      // StorageManager relies on state transition messages to add and remove replicasOnFailedDisks, if there is no state transition
+      // messages, then when we reset the partitions, replica list in the StorageManager would be obsolete. Fortunately
+      // all the replicasOnFailedDisks in the failed disks are already stopped, we just have to remove the disk from disk maps.
+      failedDisk.addAll(newFailedDisks);
+      long healthyDiskCapacity = diskToDiskManager.keySet()
+          .stream()
+          .filter(((Predicate<DiskId>) failedDisk::contains).negate())
+          .mapToLong(DiskId::getRawCapacityInBytes)
+          .sum();
+      List<ReplicaId> replicasOnFailedDisks = partitionNameToReplicaId.values()
+          .stream()
+          .filter(replica -> newFailedDisks.contains(replica.getDiskId()))
+          .collect(Collectors.toList());
+      logger.info("Replicas on the failed disk: {}", replicasOnFailedDisks);
+
+      long startTime = System.currentTimeMillis();
+      boolean success = false;
+      DistributedLock lock = primaryClusterParticipant.getDistributedLock("DISK_FAILURE", "Lock for disk failure");
+      while (true) {
```

Comment:
I think we don't need the outer while loop.
We only need while for the lock.tryLock.

Reply:
but we want to backoff a bit when we can't grab the lock

Reply:
We don't need the outer while(true). Change the lock from if to while. Inside this inner while, we have the backoff. 
In this way, we can get rid of all the continue, break and the outer while.

```
while (!lock.tryLock()) {
          // sleep for a while and try to acquire lock again
          logger.info("Fail to acquire lock when handling disk failure, backoff some time and retry");
          storeMainMetrics.handleDiskFailureRetryLockCount.inc();
          backoff();
}
Then continue all the other operations.
```

Reply:
make sense. Updated.